### DISCUSSION TITLE
Build OpenRadioss with Visual Studio in Windows

### DISCUSSION THREAD

**Discussion by xujiangping** [OpenRadioss/OpenRadioss#1475] [Q&A] [answered]
Hello, I am encountering an error when building OpenRadioss with Visual Studio in Windows (Windows 10, Visual studio 2022, oneAPI 2023.1):
Variable CMAKE_ C_ COMPILER defined multiple times
Variable CMAKE_ CXX_ COMPILER defined multiple times
There two errors repeated six times due to the six configurations (Starter_x64-Debug,Starter_x64-Release, etc..) in the output list of Visual studio IDE.
The CMakeLists and CMakeSettings are attached.
Thank you very much in advance!
[CMakeLists.txt](https://github.com/OpenRadioss/OpenRadioss/files/11484074/CMakeLists.txt)
[CMakeSettings.txt](https://github.com/OpenRadioss/OpenRadioss/files/11484093/CMakeSettings.txt)

**Comment by OlivierWienholtz** [comment]
Hi
Those variables are set in the CMakeSettings.json file to force compiler to use Intel OneAPI (C& C++)
Visual Studio tries to set its own. This is why it is set twice.

Those errors do not prevent the compilation. You can ignore those & proceed with build.

**Reply by xujiangping** [reply]
Thank you! The errors mentioned above are obtained from the building process in Visual studio. So how to ignore those errors. Could you give some details? The steps that I am following are given below: 
1 install Visual Studio Community 2022, one API Base Toolkit 2023.1 and one API HPC Toolkit 2023.1;
2 Install Git and git clone the source.
3 Launch Visual Studio and Visual Studio starts and read the configuration files CMakeLists.txt and CMakeSettings.json.
4 Select build configuration in the Configuration menu, Choose the wanted configuration from Starter/Engine release and debug configuration.
5 Launch in Menu : [Build][Build All] . Then got the errors.

**Comment by OlivierWienholtz** [comment]
In outpu you should see (here Starter Build)


![output](https://github.com/OpenRadioss/OpenRadioss/assets/98882516/d79f7c94-21c1-4ac4-810d-7522fb4d5680)


But the errors you mentioned are present : 
![vs_errors](https://github.com/OpenRadioss/OpenRadioss/assets/98882516/6549aba2-e3fe-48d7-923d-5335242910eb)

**Reply by xujiangping** [reply]
The first figure is different form mine and the second is the same. The captured screen is attached and output information of build is given below:

>------ Build All started: Project: OpenRadioss, Configuration: Starter_x64-Debug ------
  [1/3998] Generating CMakeFiles/includes_starter_win64_db/__foo.inc, CMakeFiles/includes_starter_win64_db/build_info.inc
  FAILED: starter/CMakeFiles/includes_starter_win64_db/__foo.inc starter/CMakeFiles/includes_starter_win64_db/build_info.inc C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/__foo.inc C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/build_info.inc 
  cmd.exe /C "cd /D C:\OpenRadioss\vs_build_win64\Starter_x64-Debug\starter && python C:/OpenRadioss/starter/../Compiling_tools/script/or_build_info.py -arch=win64 -outfile=C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/build_info.inc"
  Python was not found but can be installed from the Microsoft Store: ms-windows-store://pdp/?productid=9NJ46SX7X90P[2/3998] Generating CMakeFiles/includes_starter_win64_db/starter_message_description.inc
  
  FAILED: starter/CMakeFiles/includes_starter_win64_db/starter_message_description.inc C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/starter_message_description.inc 
  cmd.exe /C "cd /D C:\OpenRadioss\vs_build_win64\Starter_x64-Debug\starter && python C:/OpenRadioss/starter/../Compiling_tools/script/merge_files.py -file1=C:/OpenRadioss/starter/source/output/message/starter_message_description.txt -file2=C:/OpenRadioss/starter/../hm_cfg_files/messages/CONFIG/msg_hw_radioss_reader.txt -outfile=C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/fullstarter_message_description.txt && python C:/OpenRadioss/starter/../Compiling_tools/script/build_message.py -inputfile=C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/fullstarter_message_description.txt -outfile=C:/OpenRadioss/vs_build_win64/Starter_x64-Debug/starter/CMakeFiles/includes_starter_win64_db/starter_message_description.inc"
  Python was not found but can be installed from the Microsoft Store: ms-windows-store://pdp/?productid=9NJ46SX7X90P
  ninja: build stopped: subcommand failed.

Build All failed.

![BuildOutput](https://github.com/OpenRadioss/OpenRadioss/assets/5434580/1aa65d40-0f0f-4e6d-8216-637c0b09c26f)

**Comment by OlivierWienholtz** [comment, marked as answer]
Thank You for the screenshot. This was helpfull...

The build procedure is using Python. There is an error message of missing Python. You can modify your VS Installation and install the python from there. Otherwise you can install one from Microsoft Store as written in the message.

**Reply by xujiangping** [reply]
Great！Solved by following your suggestion. Should be noted that the package "Intel distribution for Python" must be downloaded separately since it is not included in one API Base Toolkit. Thanks for your help!

**Comment by bernoertel** [comment]
Hello Mr. Wienholz, thank you for  INTEL-MPI, all works fine now.  
But I get Errors, when I try to build Open Radios from the sources under Win11 with Visual Studio 2019.  I have the errors from above : "Variable CMAKE_ CXX_ COMPILER defined multiple time"
and additionally the following  2 Errors from the Maker : 
---------------------------------------
CMake Error at C:\Users\boertel\Desktop\TUHH\Open_Radioss\OpenRadioss-main\CMakeLists.txt:6 (project)
The CMAKE_C_COMPILER
icx.exe
is not a full path and was not found in the PATH.
Tell CMAKE where to find the compiler by setting either the environment variable "CC" or the CMake Cache entry CMAKE_C_COMPILER
to the full path to the compiler, or to the compiler name if it is in the path. 
 ---------------------------------------
Inside Visual Studio I find under  "Manage Configurations" the following Window  "CMake Settings"
 ![Screenshot 2025-02-15  #214658](https://github.com/user-attachments/assets/3eb9ca53-d009-434c-8e87-750cbe0129fa)
But  icx.exe  cannot be found anywhere. And Setting CC as globale Variable to  cl.exe , which is the C-C++ Compiler Driver helps nothing.   Build and "Rebuild All"  is not existing in the Menue of VS.  I have installed all Prerequesis except Git.
Sincerly         bernoertel

**Reply by bernoertel** [reply]
I have found compilers here  :  C:\Program Files (x86)\Intel\oneAPI\compiler\latest\bin\compiler>clang++.exe  for example.  And in another path Fortran Compiler

**Comment by bernoertel** [comment]
thank you very much, but later.


Gesendet: Montag, 17. Februar 2025 um 16:10
Von: "Olivier Wienholtz" ***@***.***>
An: OpenRadioss/OpenRadioss ***@***.***>
CC: bernoertel ***@***.***>,Mention ***@***.***>
Betreff: Re: [OpenRadioss/OpenRadioss] Build OpenRadioss with Visual Studio in Windows (Discussion #1475)

 
Hello @bernoertel
OpenRadioss uses the Intel OneAPI ifx and icx compiler.to load the compiler on your installation, you can type in cmd Shell :call "C:\Program Files (x86)\Intel\oneAPI\setvars"
This will load all variables to use the Intel Compilers.after this command, you can verify the compilers are available by typing:ifx --versionicx --version
You can find more detailed Installation instructions on:https://github.com/OpenRadioss/OpenRadioss/blob/main/HOWTO.md#build-environment-on-windows
Best Regards,Olivier

**Reply by bernoertel** [reply]
Hello Mr. Wienholtz,
I have tryed to compile and bind Open Radioss in a command windows using the script  "building_windows.bat"   with a lot of arguments,  but that is very laborious. Applying setvars is not enough, all errors from Visual Studio remain.  Compilers of VS are not found.
And what did I forget, that I get this ninja-error  in the last row  ??
![image](https://github.com/user-attachments/assets/30d30de7-5392-4f7b-a90f-af548eabf900)
But there is an advice :  **Intel OneAPI plugin for VS2022 must be installed and running.**
**I use VS2019,** I have googelt but cannot find such a plugin. 
Can you please provide me such a plugin for VS2019 or the URL, from where I can download it, 
resp. explain what this remark wants to tell me  ??
BR  bernoertel  ``

**Reply by bernoertel** [reply]
Hello Mr. Wienholtz,
Problem solved, Starter- and  Engine-Binary build with VS-Studio.  Marin Bulla showed me today. 
The command in command-Window was  :  **build_windows.bat -arch=win64_ifx -nt=4**
in starter- and engine Directory individually.
Looks simple, but for me impossible to guess.  :-)
BR
Bernhard

**Reply by OlivierWienholtz** [reply]
Hello @bernoertel 

Glad to see it is working now.
Please take care : we are currently changing build flags & defaults for Windows compilers.
This follows Intel OneAPI compiler changes.
Most important changes have already been brought in.

We have made the ifx compiler as default.
Now -arch=win64   become the default ifx base IntelOneAPI (IntelOneAPI 2025.0 or higher)
We have a fallback to 
-arch=win64_ifort for those using older OneAPI versions.

That said, you might have a source code version older than the change I am talking about.

Please note as well:
We are currently preparing some other change & defaults which might impact you but brings other defaults changes.
We will keep you & all in touch

Grüsse
Olivier Wienholtz

**Reply by bernoertel** [reply]
Hallo Hr. Wienholtz,
when compiling from the sources under Win11 , I got the following 2 exe - files after linking in the exec-directory :
![image](https://github.com/user-attachments/assets/a3c253cc-e830-40df-bc89-2b4d03bf9bcd)
When I want to check, if they are correct, against which files in the Open Radioss exec directory must I exchenge them
and how must I rename them  ??  
![image](https://github.com/user-attachments/assets/b1d3f0f9-5b53-4dcd-8007-d2be343cea5f)
Or must I choose a different way to check the correctness of my 2 new created exe-files  ??
Grüsse von boertel    (Bernhard)

**Comment by OlivierWienholtz** [comment]
Hello @bernoertel 

You executable may differ from the builds, as Compiler Version / Build platform / date / source version etc... differs.
There is a small test in OpenRadioss, you can use it to verify your build is good.

In shell loading the Intel Compiler Runtime : 
enter in qa-tests directory
invoke for a 4 threads tests : 

test_script.bat -arch=win64_ifx  -nt=4   

Hope it helps,
Best Regards
Olivier

**Reply by bernoertel** [reply]
Hallo Hr. Wienholtz,
**I got 100% tests passed, 0 tests failed out of 78** , means OK.  
But can I use these  2 .exe files now in an existing binary installation of  Open Radioss, as you see in my screenshot above  ??
BR
Bernhard
